File: cases/union-find/weighted-quick-union.lua
UF = {}
UF.__index = UF

function UF:new()
    local t ={}
    setmetatable(t, UF)
    t.cnt = 0 -- 连通分量的数量
    t.ids = {} -- 父链接数组(由触点索引)
    t.sz = {} -- (由触点索引的) 各个根节点所对应的分量的大小
    return t
end

function UF:init(n)
    self.cnt = n
    for i = 1, n do
        self.ids[i] = i
        self.sz[i] = 1
    end
end

function UF:count()
    return self.cnt
end

function UF:connected(p, q)
    return self:find(p) == self:find(q)
end

--[[ impl-3
    weighted quick-union
]]
function UF:find(p)
    -- 找出分量的名称 跟随连接找到根节点
    -- p != id[p]
    while p ~= self.ids[p]  do
        p = self.ids[p]
    end
    return p
end

function UF:union(p, q)
    -- 将 p 和 q 的根节点统一
    local i = self:find(p)
    local j = self:find(q)

    -- 将小树的根节点连接到大树的根节点
    if self.sz[i] < self.sz[j] then
        self.ids[i] = j
        self.sz[j] = self.sz[j] + self.sz[i]
    else
        self.ids[j] = i
        self.sz[i] = self.sz[i] + self.sz[j]
    end
    self.cnt = self.cnt - 1
end

local function test()
    local N = 10  -- 读取触点数量
    local uf = UF:new()
    uf:init(N)

    local pSet = {5,4,7,10,3,6,8,7}
    local qSet = {4,9,6,5,2,1,3,2}

    for i = 1, #pSet, 1 do
        if uf:connected(pSet[i], qSet[i]) then
            -- 如果已经联通 则忽略
            goto continue
        end
        uf:union(pSet[i], qSet[i]) -- 归并分量
        print(pSet[i] .. " --- " .. qSet[i])
        ::continue::
    end
    print(uf:count() .. " components")
end
test()

File: cases/union-find/weighted-quick-union.rb
class UF
    attr_reader :count

    def initialize()
        @count = 0
        @ids = []
        @sz = []
    end
    
    def init(n)
        (0...n).each do |i|
            @ids[i] = i
            @sz[i] = 1
        end
        @count = n
    end

    def connected(p, q)
        find(p) == find(q)
    end

    # weighted quick-union
    def find(p)
        while p != @ids[p]
            p = @ids[p]
        end
        return p
    end

    def union(p, q)
        # 将p, q 归并到相同分量中
        pid = find(p)
        qid = find(q)

        if pid == qid 
            return
        end

        # 小树 挂 大树
        if @sz[pid] < @sz[qid]
            @ids[pid] = qid
            @sz[qid] += @sz[pid]
        else
            @ids[qid] = pid
            @sz[pid] += @sz[qid]
        end

        @count -= 1
    end
end


def test
    pSet = [4,3,6,9,2,5,7,6]
    qSet = [3,8,5,4,1,0,2,1]

    n = 10
    uf = UF.new()
    uf.init(n)

    (0...pSet.length).each do |i|
        if !uf.connected(pSet[i], qSet[i])
            uf.union(pSet[i], qSet[i])
            puts "#{pSet[i]} --- #{qSet[i]}"
        end
    end

    puts "#{uf.count} components"
end

test
